### DISCUSSION TITLE
Error casting object? to int.

### DISCUSSION THREAD

**Discussion by AlfonsByRecognize** [dotnet/csharplang#7302] [Q&A]
We have the following code:
```
	uint noproblem = (uint) 32;
	int noproblemresult = (int) noproblem;
	
	object? test = (uint) 32;
	int result = (int) test;
```
The first 2 lines doesnt throw an error, the 2nd 2 lines throw 'Unable to cast object of type 'System.UInt32' to type 'System.Int32''. Can someone explain me why?

**Comment by CyrusNajmabadi** [comment]
The bottom example is a runtime cast.  And Because a uint is not an int, the runtime fails to cast.   

The top example, is not a runtime-cast. It is a "c# conversion", and c# has special rules on what to do here when the types are statically known.

**Reply by bernd5** [reply]
Is it really the case?
I thought `uint` to `int` conversion is simply a reinterpretation of the bytes (outside of an `checked` environment) - that is the reason why no code for that is emitted. But both conversions are runtime conversions.

The actual problem is that the runtime does not unboxing + value type conversion.

So you would need to write:
```csharp
object? test = (uint) 32;
int result = (int)(uint) test; //first unbox then convert
```

**Reply by HaloFour** [reply]
> I thought uint to int conversion is simply a reinterpretation of the bytes (outside of an `checked` environment)

That would be the C# conversion aspect of it.

**Comment by huoyaoyuan** [comment]
Casting from object to value type is [unboxing](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing). It's a different type of cast than numeric casting. C# doesn't use C++-style cast operators to distinguish different casts, but only uses C-style casts.
